dg-commons/src/main/code/Graph/Graph.scala

```diff
@@ -0,0 +1,210 @@
+  */
+class Graph[T <: DisplayableData](initialNodeValue: Option[T] = None, var isEdgeLinkTrackingOn: Boolean = false, private var _graphId: String = "", var appendSharedDisplayIdsWithNumericalSuffix: Boolean = false) {
+  def this(graphId: String) {
+    this(_graphId=graphId, appendSharedDisplayIdsWithNumericalSuffix = true)
+  }
+  def this(initialNodeValue: T, graphId: String) {
+    this(initialNodeValue=Some(initialNodeValue), _graphId=graphId)
+  }
+
+  var customSeed: Option[Long] = None
+  var customGlobalSeed: Option[Short] = None
+  val customBooleanAttributes = mutable.HashMap[String, Boolean]()
+  var userConfiguredGraphId = false
+
+  /**
+   * Get the graph ID.
+   * @return
+   */
+  def graphId: String = {
+    if (_graphId.isEmpty) _graphId = RandomHelper.randomUuid().replace('-','_')
+    _graphId
+  }
+
+  /**
+   * Set the graph ID.
+   * @param value
+   */
+  def graphId_=(value: String) = {
+    _graphId = value.replace('-','_').replace(' ','_').replace("\"","")
+    userConfiguredGraphId = true
+  }
+
+  if (_graphId != "") graphId = _graphId
+
+  /**
+   * All the root nodes in this graph (nodes with no parents, but excluding orphans)
+   */
+  val rootNodes = new mutable.ArrayBuffer[Node[T]]()
+
+  /**
+   * All the nodes in this graph, in the sequence they were added.
+   */
+  var allNodes = new mutable.ArrayBuffer[Node[T]]()
+
+  /**
+   * Used when appendSharedDisplayIdsWithNumericalSuffix is true, to ensure each node data has a unique displayable ID.
+   */
+  var nodeIdCounters = collection.immutable.HashMap[String, Int]()
+
+  /**
+   * Optional structure that can be used for retracing the steps used to create the graph.
+   */
+  var edgeLinkTrackingDescriptions = collection.immutable.List[EdgeCreationDescription[T]]()
+
+  if (initialNodeValue != None) {
+    addInitialNode(initialNodeValue.get)
+  }
+
+  /**
+   * Add the first node in this graph.
+   * @param newNodeValue
+   * @return
+   */
+  def addInitialNode(newNodeValue: T): Node[T] = {
+    assert(allNodes.size == 0 && rootNodes.size == 0)
+
+    if (isEdgeLinkTrackingOn) edgeLinkTrackingDescriptions :+= new AddInitialNodeDescription[T](newNodeValue)
+
+    addNewRootNode(newNodeValue, track=false)
+  }
+
+  /**
+   * Add a new node as a root in this graph.
+   * @param newNodeValue
+   * @param track
+   * @return
+   */
+  def addNewRootNode(newNodeValue: T, track: Boolean = true): Node[T] = {
+    val newNode = new Node[T](newNodeValue, this, allNodes.size)
+    allNodes += newNode
+    rootNodes += newNode
+
+    if (isEdgeLinkTrackingOn && track) edgeLinkTrackingDescriptions :+= new AddNewRootNodeDescription[T](newNodeValue)
+
+    newNode
+  }
+
+  /**
+   * Add a new node to the graph, but don't link it to anything.
+   * TODO: This would be better accomplished by having a separate graph with this single childNode.
```

Comment:
If that's something that you really want to do, then open a ticket on github and reference yourself in the description.